Networking Guide (#249)
FabricMC/fabric-docs: develop/networking.md

---
title: Networking
description: A general guide on networking using Fabric API.
authors:
  - daomephsta
  - dicedpixels
  - earthcomputer
  - flooferland
  - fxmorin
  - i509vcb
  - modmuss50
  - natanfudge
  - netuserget
  - nshak
  - parzivail
  - skycatminepokie
  - solidblock
  - voleil
  - wxffel
  - ytg123-mods

>   - zulrang

Reviewer: ```suggestion
authors:
  - Daomephsta
  - dicedpixels
  - Earthcomputer
  - FlooferLand
  - FxMorin
  - i509VCB
  - modmuss50
  - natanfudge
  - NetUserGet
  - NShak
  - parzivail
  - skycatminepokie
  - SolidBlock-cn
  - Voleil
  - Wxffel
  - YTG123-Mods
  - zulrang
```

friend, you boot up a dedicated server and invite your friend on with the mod installed.

You use the item and the server crashes. You will probably notice in the crash log an error similar to this:

```log
[Server thread/FATAL]: Error executing task on Server
java.lang.RuntimeException: Cannot load class net.minecraft.client.MinecraftClient in environment type SERVER
```

### Why Does the Server Crash? {#why-does-the-server-crash}

The code calls logic only present on the client distribution of the Minecraft. The reason for Mojang distributing the
game in this way is to cut down on the size of the Minecraft Server JAR file. There isn't really a reason to include an
entire rendering engine when your own machine will render the world.

In a development environment, client-only classes are indicated by the `@Environment(EnvType.CLIENT)` annotation.

### How Do We Fix the Crash? {#how-do-we-fix-the-crash}

To fix this issue, you need to understand how the game client and dedicated server communicate:

![Sides](/assets/develop/networking/sides.png)

The diagram above shows that the game client and dedicated server are separate systems, bridged together using

> _packets_. Packets can contain data which we refer to as the _payload_.

Reviewer 2: This is fine I guess, it just doenst have anything to do with networking. Packets are still sent between the logicial client and server even in single player, or when on a LAN world. A better example would be to cause desync by setting something only the logicial client. Maybe try adding a block to the hotbar from a keybind or something.

your game client can't directly access the server instance.

## An Introduction to Networking {#an-introduction-to-networking}

### Defining a Payload {#defining-a-payload}

This can be done by creating a Java `Record` with a `BlockPos` parameter that implements `CustomPayload`.

@[code lang=java transcludeWith=:::summon_Lightning_payload](@/reference/latest/src/main/java/com/example/docs/networking/basic/SummonLightningS2CPayload.java)

At the same time, we've defined:

- An `Identifier` used to identify our packet's payload. For this example our identifier will be
  `fabric-docs-reference:summon_lightning`.

@[code lang=java transclude={13-13}](@/reference/latest/src/main/java/com/example/docs/networking/basic/SummonLightningS2CPayload.java)

- A public static instance of `CustomPayload.Id` to uniquely identify this custom payload. We will be referencing this
  ID in both our common and client code.

@[code lang=java transclude={14-14}](@/reference/latest/src/main/java/com/example/docs/networking/basic/SummonLightningS2CPayload.java)

- A public static instance of a `PacketCodec` so that the game knows how to serialize/deserialize the contents of the
  packet.

> @[code lang=java transclude={15-15}](@/reference/latest/src/main/java/com/example/docs/networking/basic/SummonLightningS2CPayload.java)

Reviewer 3: ![image](https://github.com/user-attachments/assets/296e1a1e-6965-4e44-8fe0-b2c7b0dd89b4)

Author: My bad, I have IJ configured to sort fields 😅. Fixed.

Reviewer: out of curiosity, how does your IDE sort them? alphabetically?

We have also overridden `getId` to return our payload ID.

@[code lang=java transclude={17-20}](@/reference/latest/src/main/java/com/example/docs/networking/basic/SummonLightningS2CPayload.java)

### Registering a Payload {#registering-a-payload}

Before we send a packet with our custom payload, we need to register it.

::: info
`S2C` and `C2S` are two common suffixes that mean _Server-to-Client_ and _Client-to-Server_ respectively.
:::

> This can be done in our **common initializer** by `PayloadTypeRegistry.playS2C().register` which takes in a

Reviewer 3: ```suggestion
This can be done in our **common initializer** by using `PayloadTypeRegistry.playS2C().register`, which takes in a
```

`CustomPayload.Id` and a `PacketCodec`.

@[code lang=java transclude={25-25}](@/reference/latest/src/main/java/com/example/docs/networking/basic/FabricDocsReferenceNetworkingBasic.java)

A similar method exists to register client-to-server payloads: `PayloadTypeRegistry.playC2S().register`.

### Sending a Packet to the Client {#sending-a-packet-to-the-client}

To send a packet with our custom payload, we can use `ServerPlayNetworking.send` which takes in a `ServerPlayerEntity`
and a `CustomPayload`.

> Let's start by creating our Lightning Tater item. You can override `use` to add trigger an action when the item is used.

Reviewer 3: ```suggestion
Let's start by creating our Lightning Tater item. You can override `use` to trigger an action when the item is used.
```

> In this case, let's send packets to all the players in the server world.

Reviewer 3: ```suggestion
In this case, let's send packets to all players in the server world.
```

@[code lang=java transcludeWith=:::lightning_tater_item](@/reference/latest/src/main/java/com/example/docs/networking/basic/LightningTaterItem.java)

Let's examine the code above.

We only send packets when the action is initiated on the server, by returning early with a `isClient` check:

@[code lang=java transclude={22-24}](@/reference/latest/src/main/java/com/example/docs/networking/basic/LightningTaterItem.java)

We create an instance of the payload with the user's position:

@[code lang=java transclude={26-26}](@/reference/latest/src/main/java/com/example/docs/networking/basic/LightningTaterItem.java)

> Finally, we get all the players in the server world through `PlayerLookup` and send a packet to each player.

Reviewer 3: ```suggestion
Finally, we get all players in the server world through `PlayerLookup` and send a packet to each player.
```

Then we register our payload in our **common initializer**. However, this time as _Client-to-Server_ payload by using
`PayloadTypeRegistry.playC2S().register`.

@[code lang=java transclude={26-26}](@/reference/latest/src/main/java/com/example/docs/networking/basic/FabricDocsReferenceNetworkingBasic.java)

To send a packet, let's add an action when the player uses a Poisonous Potato. We'll be using the `UseEntityCallback`
event to
keep things concise.

We register the event in our **client initializer**, and we use `isClient()` to ensure that the action is only triggered
on the logical client.

@[code lang=java transcludeWith=:::use_entity_callback](@/reference/latest/src/client/java/com/example/docs/network/basic/FabricDocsReferenceNetworkingBasicClient.java)

We create an instance of our `GiveGlowingEffectC2SPayload` with the necessary arguments. In this case, the network ID
of
the targeted entity.

@[code lang=java transclude={50-50}](@/reference/latest/src/client/java/com/example/docs/network/basic/FabricDocsReferenceNetworkingBasicClient.java)

Finally, we send a packet to the server by calling `ClientPlayNetworking.send` with the instance of our

> `UsePoisonousPotatoPayload`.

Reviewer 3: ```suggestion
`GiveGlowingEffectC2SPayload`.
```

  packet.

  @[code lang=java transclude={11-11}](@/reference/latest/src/main/java/com/example/docs/networking/basic/SummonLightningPayload.java)

We have also overridden `getId` to return our payload ID.

### Registering a Payload {#registering-a-payload}

Before we send a packet with our custom payload, we need to register it.

This can be done in our **common initializer** by `PayloadTypeRegistry.playS2C().register` which takes in a
`CustomPayload.Id` and a `PacketCodec`.

@[code lang=java transclude={22-22}](@/reference/latest/src/main/java/com/example/docs/networking/basic/FabricDocsReferenceNetworkingBasic.java)

A similar method exists to register client-to-server payloads: `PayloadTypeRegistry.playC2S().register`.

`S2C` and `C2S` are two common suffixes that mean _Server-to-Client_ and _Client-to-Server_ respectively.

### Sending a Packet to the Client {#sending-a-packet-to-the-client}

To send a packet with our custom payload, we can use `ServerPlayNetworking.send` which takes in a `ServerPlayerEntity`
and a `CustomPayload`.

Let's start by creating our Lightning Tater item. You can override `use` to add trigger an action when the item is used.

> In this case, let's send packets to all the players in the server.

Reviewer 3: This statement is wrong. In the example code, the payload is only sent to players tracking the `BlockPos`. Tracking is a very important concept for efficient networking, it shouldn't be omitted.

Author: Noted. I believe using `world` would be better to match the description. Then I could add a statement about tracking and why it is important.

Author: I've now changed the example to match what's written and added an info block about `PlayerLookup`.

@[code lang=java transcludeWith=:::lightning_tater_item](@/reference/latest/src/main/java/com/example/docs/networking/basic/LightningTaterItem.java)

Let's examine the code above.

We only send packets when the action is initiated on the server, by returning early with a `isClient` check:

@[code lang=java transclude={23-25}](@/reference/latest/src/main/java/com/example/docs/networking/basic/LightningTaterItem.java)

We create an instance of the payload with the user's position:

@[code lang=java transclude={27-27}](@/reference/latest/src/main/java/com/example/docs/networking/basic/LightningTaterItem.java)

> Finally, we get all the players in the server through `PlayerLookup` and send a packet to each player.

Reviewer 3: Ditto

Author: Addressed through previous comment.

can be used to get the `x`, `y` and `z` positions.

@[code lang=java transclude={27-27}](@/reference/latest/src/client/java/com/example/docs/network/basic/FabricDocsReferenceNetworkingBasicClient.java)

Finally, we create a `LightningEntity` and add it to the world.

@[code lang=java transclude={29-34}](@/reference/latest/src/client/java/com/example/docs/network/basic/FabricDocsReferenceNetworkingBasicClient.java)

Now, if you add this mod to a server and when a player uses our Lightning Tater item, every player will see lightning
striking at the user's position.

<VideoPlayer src="/assets/develop/networking/summon-lightning.webm" title="Summon lightning using Lightning Tater" />

### Sending a Packet to the Server {#sending-a-packet-to-the-server}

Just like sending a packet to the client, we start by creating a custom payload. This time, we will be sending the
player's name and position to the server.

@[code lang=java transcludeWith=::use_poisonous_potato_payload](@/reference/latest/src/main/java/com/example/docs/networking/basic/UsePoisonousPotatoPayload.java)

> Note the usage of `PacketCodec.tuple` to combine two values, a `String` and a `BlockPos`.

Reviewer 3: I know this is done for the sake of example, but this example is inefficient and dangerous:
1) Sending any data about the player is pointless because the server already knows which player sent it and can access any information it needs there.
2) If the client or server want to refer to another entity in a packet, they should use it's numerical ID (not UUID)
3) A malicious client can exploit this because the server doesn't check that any data is factually correct (data validation in C2S packets is also an important topic).

Author: 1. Yes, this I figured. I wanted a simple example.
2. I believe we can overlook this for the sake of simplicity.
3. This however is concerning. I was not aware and didn't think that far as we're dealing with the basics. I'm open for a better example that could address above 2 points as well.

Reviewer 3: If I were to write this, I would add a situation where a player is casting some sort of spell to another entity. In this situation, efficient networking can be showcased (see 2) and proper data validation can be showcased with checks like "does the entity actually exist and is not invulnerable?" "is the player close enough to the entity?"

Author: I've replaced the example with a simple instance of applying an effect on to a living entity, including some checks. Also added note on the importance of server side validation. Hopefully, it's enough to cover the points raised here while keeping it simple for the docs.

We pass in the appropriate codec along with a method reference to get the value from the Record to build this compound
codec.

`USE_POISONOUS_POTATO_PAYLOAD_ID` is defined in the **common initializer** for easy access.

@[code lang=java transclude={17-17}](@/reference/latest/src/main/java/com/example/docs/networking/basic/FabricDocsReferenceNetworkingBasic.java)

Then we register our payload in our **common initializer**. However, this time as _Client-to-Server_ payload by using
`PayloadTypeRegistry.playC2S().register`.

@[code lang=java transclude={22-22}](@/reference/latest/src/main/java/com/example/docs/networking/basic/FabricDocsReferenceNetworkingBasic.java)

To send a packet, let's add an action when the player uses a Poisonous Potato. We'll be using the `UseItemCallback`
event to
keep things concise.

> We register the event in our **client initializer**. This ensures that the action is only triggered on the client.

Reviewer 3: This isn't how sides work. A callback registered on the physical client can still be called on the logical server (when the client is an integrated server host)

Author: ~~I'll remove this statement and move the callback registration to the logical side.~~

Edit: I don't think I understood this correctly. Any suggestions to rewrite this better?

Author: I believe adding a `isClient` check would restrict the behavior to the logical client. I've made this change. Hopefully, this is a better way of writing this example. If not, guide me through it.

---
title: Networking
description: A general guide on networking using Fabric API.
authors:
  - daomephsta
  - dicedpixels
  - earthcomputer
  - flooferland
  - fxmorin
  - i509vcb
  - modmuss50
  - natanfudge
  - netuserget
  - nshak
  - parzivail
  - skycatminepokie
  - solidblock
  - voleil
  - wxffel
  - ytg123-mods
  - zulrang
---

Networking in Minecraft is used so the client and server can communicate with each other. Networking is a broad topic,
so this page is split up into a few categories.

## Why Is Networking Important? {#why-is-networking-important}

The importance of networking can be shown by a simple code example.

::: warning
Below code is for demonstration purposes only.
:::

Say you had a Wand which highlights the block you're looking, which will be visible to all
nearby players:

```java

> public class HighlightingWandItem extends Item {

Reviewer 4: Needs to be in the reference mod in some way because of the `use` method, this will most certainly change in the future.

Author: Given this is just a "bad example", I'd rather have it here than in the mod. I'd expect it'll be a while before Items would change.

However, I'm open to suggestions.

Reviewer 3: Maybe we could re-use a vanilla item again (blaze rod?), like we did with the Poisonous Potato

Author: Yes, I'm just reluctant to add any code that would crash the mod 😅.

Reviewer: That's exactly the reason why we need the code in the mod

Author: Then I'll add a simpler example like adding a particle. Block highlighting sounds complex to implement. I'm open to other bad examples as well.

Author: Now that I'm looking into this, how am I supposed to call a client class on the server when we're using split sources? I'd say we leave the example and perhaps add an info box highlighting how this wouldn't happen on a split source project.

Reviewer 2: You dont, as that would crash the game, split sources is correctly stopping you from doing it. With packets you would register the client side handler in the client source set.

Author: Yep, we can't demonstrate the bad example. I don't think it's necessary to include it in the reference mod. Even if the item's `use` method changes, it would be just a small piece of code to update.

        return super.use(world, user, hand);
    }
}
```

Upon testing, you will see a lightning bolt being summoned and nothing crashes. Now you want to show the mod to your
friend, you boot up a dedicated server and invite your friend on with the mod installed.

You use the item and the server crashes. You will probably notice in the crash log an error similar to this:

```log
[Server thread/FATAL]: Error executing task on Server
java.lang.RuntimeException: Cannot load class net.minecraft.client.MinecraftClient in environment type SERVER
```

### Why Does the Server Crash? {#why-does-the-server-crash}

The code calls logic only present on the client distribution of the Minecraft. The reason for Mojang distributing the
game in this way is to cut down on the size of the Minecraft Server JAR file. There isn't really a reason to include an
entire rendering engine when your own machine will render the world.

In a development environment, client-only classes are indicated by the `@Environment(EnvType.CLIENT)` annotation.

### How Do We Fix the Crash? {#how-do-we-fix-the-crash}

To fix this issue, you need to understand how Minecraft communicates between the game client and dedicated

> server:

Reviewer 3: ```suggestion
To fix this issue, you need to understand how the game client and dedicated
server communicate:
```

_packets_. Packets can contain data which we refer to as the _payload_.

This packet bridge does not only exist between a game client and dedicated server, but also between your client and
another client connected over LAN. The packet bridge is also present even in singleplayer. This is because the game
client will spin up a special integrated server instance to run the game on.

The key difference between the three types of connections that are shown in the table below:

| Connection Type               | Access to Game Client      |
|-------------------------------|----------------------------|
| Connected to dedicated Server | None (Server crash)        |
| Connected over LAN            | Yes (Not host game client) |
| singleplayer (or LAN host)    | Yes (Full access)          |

It may seem complicated to have communication with the server in three different ways. However, you don't need to
communicate in three different ways with the game client. Since all three connection types communicate with the game
client using packets, you only need to communicate with the game client like you're always running on a dedicated

> server.

Reviewer 3: ```suggestion
```
The relation between "connection type" and "access to game client" is logically incorrect, since it is the client that connects to the server.
Access to the game client depends on the type of server: integrated or dedicated, but in the context of networking it doesn't matter

sending packets.
:::

### Receiving a Packet on the Client {#receiving-a-packet-on-the-client}

To receive a packet sent from a server on the client, you need to specify how you will handle the incoming packet.

This can be done in the **client initializer**, by calling `ClientPlayNetworking.registerGlobalReceiver` and passing a
`CustomPayload.Id` and a `PlayPayloadHandler`, which is a Functional Interface.

In this case, we'll define the action to trigger within the implementation of `PlayPayloadHandler` implementation (as a
lambda expression).

@[code lang=java transcludeWith=:::client_global_receiver](@/reference/latest/src/client/java/com/example/docs/network/basic/FabricDocsReferenceNetworkingBasicClient.java)

Let's examine the code above.

We can access the data from our payload by calling the Record's getter methods. In this case `payload.pos()`. Which then
can be used to get the `x`, `y` and `z` positions.

@[code lang=java transclude={37-37}](@/reference/latest/src/client/java/com/example/docs/network/basic/FabricDocsReferenceNetworkingBasicClient.java)

Finally, we create a `LightningEntity` and add it to the world.

> @[code lang=java transclude={38-43}](@/reference/latest/src/client/java/com/example/docs/network/basic/FabricDocsReferenceNetworkingBasicClient.java)

Reviewer 3: ![image](https://github.com/user-attachments/assets/5bfca5f2-2e9b-4183-91e7-dec5b39fcd59)

Poisonous Potato on a living entity, we request the server to apply the Glowing effect to it.

@[code lang=java transcludeWith=:::give_glowing_effect_payload](@/reference/latest/src/main/java/com/example/docs/networking/basic/GiveGlowingEffectC2SPayload.java)

We pass in the appropriate codec along with a method reference to get the value from the Record to build this codec.

Then we register our payload in our **common initializer**. However, this time as _Client-to-Server_ payload by using
`PayloadTypeRegistry.playC2S().register`.

@[code lang=java transclude={26-26}](@/reference/latest/src/main/java/com/example/docs/networking/basic/FabricDocsReferenceNetworkingBasic.java)

To send a packet, let's add an action when the player uses a Poisonous Potato. We'll be using the `UseEntityCallback`
event to
keep things concise.

We register the event in our **client initializer**, and we use `isClient()` to ensure that the action is only triggered
on the logical client.

@[code lang=java transcludeWith=:::use_entity_callback](@/reference/latest/src/client/java/com/example/docs/network/basic/FabricDocsReferenceNetworkingBasicClient.java)

> We create and instance of our `GiveGlowingEffectC2SPayload` with the necessary arguments. In this case, the network ID

Reviewer 3: ```suggestion
We create an instance of our `GiveGlowingEffectC2SPayload` with the necessary arguments. In this case, the network ID
```

of
the targeted entity.

@[code lang=java transclude={56-56}](@/reference/latest/src/client/java/com/example/docs/network/basic/FabricDocsReferenceNetworkingBasicClient.java)

Finally, we send a packet to the server by calling `ClientPlayNetworking.send` with the instance of our
`UsePoisonousPotatoPayload`.

> @[code lang=java transclude={57-57}](@/reference/latest/src/client/java/com/example/docs/network/basic/FabricDocsReferenceNetworkingBasicClient.java)

Reviewer 3: ![image](https://github.com/user-attachments/assets/52e063a9-daf4-4ed8-8692-061c10cf1da5)

player's name and position to the server.

@[code lang=java transcludeWith=::use_poisonous_potato_payload](@/reference/latest/src/main/java/com/example/docs/networking/basic/UsePoisonousPotatoPayload.java)

Note the usage of `PacketCodec.tuple` to combine two values, a `String` and a `BlockPos`.

We pass in the appropriate codec along with a method reference to get the value from the Record to build this compound
codec.

`USE_POISONOUS_POTATO_PAYLOAD_ID` is defined in the **common initializer** for easy access.

@[code lang=java transclude={17-17}](@/reference/latest/src/main/java/com/example/docs/networking/basic/FabricDocsReferenceNetworkingBasic.java)

Then we register our payload in our **common initializer**. However, this time as _Client-to-Server_ payload by using
`PayloadTypeRegistry.playC2S().register`.

@[code lang=java transclude={22-22}](@/reference/latest/src/main/java/com/example/docs/networking/basic/FabricDocsReferenceNetworkingBasic.java)

To send a packet, let's add an action when the player uses a Poisonous Potato. We'll be using the `UseItemCallback`
event to
keep things concise.

> We register the event in our **client initializer**. This ensures that the action is only triggered on the client.

Reviewer 3: ```suggestion
We register the event in our **client initializer** and we use `isClient()` to ensure that the action is only triggered on the logical client.
```

---
title: Networking
description: A general guide on networking using Fabric API.
authors:
  - dicedpixels
  - FlooferLand
  - skycatminepokie
  - fxmorin
  - netuserget
  - wxffel
  - daomephsta
  - solidblock
  - voleil
  - daomephsta
  - YTG123-Mods
  - zulrang
  - i509VCB
  - nshak
  - earthcomputer
  - natanfudge
  - modmuss50
---

Networking in Minecraft is used so the client and server can communicate with each other. Networking is a broad topic,
so this page is split up into a few categories.

## Why Is Networking Important? {#why-is-networking-important}

The importance of networking can be shown by a simple code example.

::: warning
Below code is for demonstration purposes only.
:::

Say you had a Wand which highlights the block you're looking, which will be visible to all
nearby players:

```java
public class HighlightingWandItem extends Item {
    public HighlightingWandItem(Item.Settings settings) {

>         super(settings)

Reviewer: ```suggestion
        super(settings);
```

    }

    public TypedActionResult<ItemStack> use(World world, PlayerEntity user, Hand hand) {

>         // Raycast and find the block the user is facing at
>         BlockPos target = ...

Reviewer: ```suggestion
        BlockPos target = ...
```

let's not give instructions on how to use this code (:

>         // BAD CODE: DON'T EVER DO THIS!
>         ClientBlockHighlighting.highlightBlock(MinecraftClient.getInstance(), target);

Reviewer: ```suggestion
        // BAD CODE: DON'T EVER DO THIS! // [!code error]
        ClientBlockHighlighting.highlightBlock(MinecraftClient.getInstance(), target); // [!code error]
```

        return super.use(world, user, hand);
    }
}
```

Upon testing, you will see a lightning bolt being summoned and nothing crashes. Now you want to show the mod to your
friend, you boot up a dedicated server and invite your friend on with the mod installed.

You use the item and the server crashes. You will probably notice in the crash log an error similar to this:

> ```text

Reviewer: ```suggestion
```log
```

[Server thread/FATAL]: Error executing task on Server
java.lang.RuntimeException: Cannot load class net.minecraft.client.MinecraftClient in environment type SERVER
```

### Why Does the Server Crash? {#why-does-the-server-crash}

The code calls logic only present on the client distribution of the Minecraft. The reason for Mojang distributing the
game in this way is to cut down on the size of the Minecraft Server JAR file. There isn't really a reason to include an
entire rendering engine when your own machine will render the world.

In a development environment, client-only classes are indicated by the `@Environment(EnvType.CLIENT)` annotation.

### How Do We Fix the Crash? {#how-do-we-fix-the-crash}

To fix this issue, you need to understand how Minecraft communicates between the game client and dedicated
server.

![Sides](/assets/develop/networking/sides.png)

The diagram above shows that the game client and dedicated server are separate systems, bridged together using
_packets_. Packets can contain data which we refer to as the _payload_.

This packet bridge does not only exist between a game client and dedicated server, but also between your client and

> another client connected over LAN. The packet bridge is also present even in single-player. This is because the game

Reviewer: ```suggestion
another client connected over LAN. The packet bridge is also present even in singleplayer. This is because the game
```

client will spin up a special integrated server instance to run the game on.

The key difference between the three types of connections that are shown in the table below:

| Connection Type               | Access to Game Client      |
|-------------------------------|----------------------------|
| Connected to dedicated Server | None (Server crash)        |
| Connected over LAN            | Yes (Not host game client) |

> | Single-player (or LAN host)   | Yes (Full access)          |

Reviewer: ```suggestion
| Singleplayer (or LAN host)    | Yes (Full access)          |
```

It may seem complicated to have communication with the server in three different ways. However, you don't need to
communicate in three different ways with the game client. Since all three connection types communicate with the game
client using packets, you only need to communicate with the game client like you're always running on a dedicated
server.

> Connection to a server over LAN or single-player can also be treated like the server is a remote, dedicated server; so

Reviewer: ```suggestion
Connection to a server over LAN or singleplayer can also be treated like the server is a remote, dedicated server; so
```

At the same time, we've defined:

- A public static instance of `CustomPayload.Id` to uniquely identify this custom payload. We will be referencing this
  ID in both our common and client code.

  @[code lang=java transclude={10-10}](@/reference/latest/src/main/java/com/example/docs/networking/basic/SummonLightningPayload.java)

- A public static instance of a `PacketCodec` so that the game knows how to serialize/deserialize the contents of the
  packet.

  @[code lang=java transclude={11-11}](@/reference/latest/src/main/java/com/example/docs/networking/basic/SummonLightningPayload.java)

We have also overridden `getId` to return our payload ID.

### Registering a Payload {#registering-a-payload}

Before we send a packet with our custom payload, we need to register it.

This can be done in our **common initializer** by `PayloadTypeRegistry.playS2C().register` which takes in a
`CustomPayload.Id` and a `PacketCodec`.

@[code lang=java transclude={28-28}](@/reference/latest/src/main/java/com/example/docs/networking/basic/FabricDocsReferenceNetworkingBasic.java)

A similar method exists to register client-to-server payloads: `PayloadTypeRegistry.playC2S().register`.

> `S2C` and `C2S` are two common suffixes that mean _Server-to-Client_ and _Client-to-Server_ respectively.

Reviewer: should this appear earlier?

Author: I moved this to an info box just after "Defining a Payload" since from that point onwards there will be several usages of S2C and C2S methods.

---
title: Networking
description: A general guide on networking using Fabric API.

> authors:
>   - dicedpixels
>   - FlooferLand
>   - skycatminepokie
>   - fxmorin
>   - netuserget
>   - wxffel
>   - daomephsta
>   - solidblock
>   - voleil
>   - daomephsta
>   - YTG123-Mods
>   - zulrang
>   - i509VCB
>   - nshak
>   - earthcomputer
>   - natanfudge
>   - modmuss50

Reviewer: added `parzivail` for the diagram, lowercased, alphabetized

```suggestion
authors:
  - daomephsta
  - dicedpixels
  - earthcomputer
  - flooferland
  - fxmorin
  - i509vcb
  - modmuss50
  - natanfudge
  - netuserget
  - nshak
  - parzivail
  - skycatminepokie
  - solidblock
  - voleil
  - wxffel
  - ytg123-mods
  - zulrang
```

        return super.use(world, user, hand);
    }
}
```

Upon testing, you will see a lightning bolt being summoned and nothing crashes. Now you want to show the mod to your
friend, you boot up a dedicated server and invite your friend on with the mod installed.

You use the item and the server crashes. You will probably notice in the crash log an error similar to this:

```log
[Server thread/FATAL]: Error executing task on Server
java.lang.RuntimeException: Cannot load class net.minecraft.client.MinecraftClient in environment type SERVER
```

### Why Does the Server Crash? {#why-does-the-server-crash}

The code calls logic only present on the client distribution of the Minecraft. The reason for Mojang distributing the
game in this way is to cut down on the size of the Minecraft Server JAR file. There isn't really a reason to include an
entire rendering engine when your own machine will render the world.

In a development environment, client-only classes are indicated by the `@Environment(EnvType.CLIENT)` annotation.

### How Do We Fix the Crash? {#how-do-we-fix-the-crash}

To fix this issue, you need to understand how Minecraft communicates between the game client and dedicated

> server.

Reviewer: ```suggestion
server:
```

| Connection Type               | Access to Game Client      |
|-------------------------------|----------------------------|
| Connected to dedicated Server | None (Server crash)        |
| Connected over LAN            | Yes (Not host game client) |
| singleplayer (or LAN host)    | Yes (Full access)          |

It may seem complicated to have communication with the server in three different ways. However, you don't need to
communicate in three different ways with the game client. Since all three connection types communicate with the game
client using packets, you only need to communicate with the game client like you're always running on a dedicated
server.

Connection to a server over LAN or singleplayer can also be treated like the server is a remote, dedicated server; so
your game client can't directly access the server instance.

## An Introduction to Networking {#an-introduction-to-networking}

### Defining a Payload {#defining-a-payload}

First, you need to define an `Identifier` used to identify our packet's payload. For this example our identifier will be
`fabric-docs-reference:summon_lightning`.

> We can define this identifier in our **common initializer** for easy access.

Reviewer 3: It would be actually easier to put it in the payload class itself

Author: Yep. Makes sense to keep them closer to where it's used. Changed.

Now, let's define the payload that will be sent with the packet. This payload will contain the position (a `BlockPos`)
of the item user.

This can be done by creating a Java `Record` with a `BlockPos` parameter that implements `CustomPayload`.

@[code lang=java transcludeWith=:::summon_Lightning_payload](@/reference/latest/src/main/java/com/example/docs/networking/basic/SummonLightningPayload.java)

At the same time, we've defined:

- A public static instance of `CustomPayload.Id` to uniquely identify this custom payload. We will be referencing this
  ID in both our common and client code.

  @[code lang=java transclude={10-10}](@/reference/latest/src/main/java/com/example/docs/networking/basic/SummonLightningPayload.java)

- A public static instance of a `PacketCodec` so that the game knows how to serialize/deserialize the contents of the
  packet.

  @[code lang=java transclude={11-11}](@/reference/latest/src/main/java/com/example/docs/networking/basic/SummonLightningPayload.java)

We have also overridden `getId` to return our payload ID.

### Registering a Payload {#registering-a-payload}

Before we send a packet with our custom payload, we need to register it.

::: info

> `S2C` and `C2S` are two common suffixes that mean _Server-to-Client_ and _Client-to-Server_ respectively.

Reviewer 3: Given the presence of this note, we could be a good example for devs and rename the payloads to `SummonLightningS2CPayload` and `GiveGlowingEffectC2SPayload` respectively?

In this case, let's send packets to all the players in the server world.

@[code lang=java transcludeWith=:::lightning_tater_item](@/reference/latest/src/main/java/com/example/docs/networking/basic/LightningTaterItem.java)

Let's examine the code above.

We only send packets when the action is initiated on the server, by returning early with a `isClient` check:

@[code lang=java transclude={22-24}](@/reference/latest/src/main/java/com/example/docs/networking/basic/LightningTaterItem.java)

We create an instance of the payload with the user's position:

@[code lang=java transclude={26-26}](@/reference/latest/src/main/java/com/example/docs/networking/basic/LightningTaterItem.java)

Finally, we get all the players in the server world through `PlayerLookup` and send a packet to each player.

@[code lang=java transclude={28-30}](@/reference/latest/src/main/java/com/example/docs/networking/basic/LightningTaterItem.java)

::: info
Fabric API provides `PlayerLookup`, a collection of helper functions that will look up players in a server.

A term frequently used to describe the functionality of these methods is "_tracking_". It means that an entity or a
chunk

> on the server is known to a player's client (within in view distance) and the entity or block entity should notify

Reviewer 3: "within their view distance" or "in their view distance"

tracking clients of changes.

Tracking is an important concept for efficient networking, so that only the necessary players are notified of changes by
sending packets.
:::

Remember to register the item in our **common initializer**.

> @[code lang=java transclude={24-25}](@/reference/latest/src/main/java/com/example/docs/networking/basic/FabricDocsReferenceNetworkingBasic.java)

Reviewer 3: Seems out of place and not relevant to this guide?

Author: Yes. Let's keep the focus on networking. There's a whole guide if someone needs clarification about registering items.

@[code lang=java transcludeWith=:::client_global_receiver](@/reference/latest/src/client/java/com/example/docs/network/basic/FabricDocsReferenceNetworkingBasicClient.java)

Let's examine the code above.

We can access the data from our payload by calling the Record's getter methods. In this case `payload.pos()`. Which then
can be used to get the `x`, `y` and `z` positions.

@[code lang=java transclude={32-32}](@/reference/latest/src/client/java/com/example/docs/network/basic/FabricDocsReferenceNetworkingBasicClient.java)

Finally, we create a `LightningEntity` and add it to the world.

@[code lang=java transclude={34-39}](@/reference/latest/src/client/java/com/example/docs/network/basic/FabricDocsReferenceNetworkingBasicClient.java)

Now, if you add this mod to a server and when a player uses our Lightning Tater item, every player will see lightning
striking at the user's position.

<VideoPlayer src="/assets/develop/networking/summon-lightning.webm" title="Summon lightning using Lightning Tater" />

### Sending a Packet to the Server {#sending-a-packet-to-the-server}

Just like sending a packet to the client, we start by creating a custom payload. This time, when a player uses a

> Poisonous Potato on a living entity, we will apply the Glowing effect to it.

Reviewer 3: A bit nitpicky and rather subjective: I think the concept of server authority needs to be ingrained into the readers. What we're doing really here is "asking the server to apply the Glowing effect to it" or "requesting that the server applies the Glowing effect"

@[code lang=java transcludeWith=:::give_glowing_effect_payload](@/reference/latest/src/main/java/com/example/docs/networking/basic/GiveGlowingEffectPayload.java)

We pass in the appropriate codec along with a method reference to get the value from the Record to build this codec.

> `GIVE_GLOWING_EFFECT_PAYLOAD_ID` is defined in the **common initializer** for easy access.

Reviewer 3: ditto
